assemblies/assembly_architect-decision-services/states-con.adoc

```diff
@@ -0,0 +1,345 @@
+
+The following rule turns the sprinkler off as soon as the fire in that room is extinguished.
+
+[source]
+----
+rule "When the fire is gone turn off the sprinkler"
+when
+    $room : Room( )
+    $sprinkler : Sprinkler( room == $room, on == true )
+    not Fire( room == $room )
+then
+    modify( $sprinkler ) { setOn( false ) };
+    System.out.println( "Turn off the sprinkler for room " + $room.getName() );
+end
+----
+
+While there is one sprinkler per room, there is just a single alarm for the building. An `Alarm` object is created when a fire occurs, but only one `Alarm` is needed for the entire building, no matter how many fires occur. Previously `not` was introduced to match the absence of a fact; now we use its complement `exists` which matches for one or more instances of some category.
+
+[source]
+----
+rule "Raise the alarm when we have one or more fires"
+when
+    exists Fire()
+then
+    insert( new Alarm() );
+    System.out.println( "Raise the alarm" );
+end
+----
+
+Likewise, when there are no fires we want to remove the alarm, so the `not` keyword can be used again.
+
+[source]
+----
+rule "Cancel the alarm when all the fires have gone"
+when
+    not Fire()
+    $alarm : Alarm()
+then
+    delete( $alarm );
+    System.out.println( "Cancel the alarm" );
+end
+----
+
+Finally there is a general health status message that is printed when the application first starts and after the alarm is removed and all sprinklers are turned off.
+
+[source]
+----
+rule "Status output when things are ok"
+when
+    not Alarm()
+    not Sprinkler( on == true )
+then
+    System.out.println( "Everything is ok" );
+end
+----
+
+As we did in the stateless session example, the above rules should be placed in a single DRL file and saved into the resources folder of your Maven project or one of its subfolders. We can then obtain a `KieSession` from the ``KieContainer``.
+The only difference is that this time we create a stateful session, whereas before we created a stateless session.
+
+[source,java]
+----
+KieServices kieServices = KieServices.Factory.get();
+KieContainer kContainer = kieServices.getKieClasspathContainer();
+KieSession ksession = kContainer.newKieSession();
+----
+
+
+With the session created you can now iterate your work with it. Four `Room` objects are created and inserted, as well as one `Sprinkler` object for each room. At this point the engine has done all of its matching, but no rules have fired yet.
+Calling `ksession.fireAllRules()` allows the matched rules to fire, but without a fire that will just produce the health message.
+
+[source,java]
+----
+String[] names = new String[]{"kitchen", "bedroom", "office", "livingroom"};
+Map<String,Room> name2room = new HashMap<String,Room>();
+for( String name: names ){
+    Room room = new Room( name );
+    name2room.put( name, room );
+    ksession.insert( room );
+    Sprinkler sprinkler = new Sprinkler( room );
+    ksession.insert( sprinkler );
+}
+
+ksession.fireAllRules();
```

Comment:
Could we please update, same reason as above:
```java
String[] names = new String[] {"kitchen", "bedroom", "office", "livingroom"};
Map<String, Room> name2room = new HashMap<String, Room>();
for (String name: names) {
    Room room = new Room(name);
    name2room.put(name, room);
    ksession.insert(room);
    Sprinkler sprinkler = new Sprinkler(room);
    ksession.insert(sprinkler);
}

ksession.fireAllRules();
```

Reply:
Done